File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/EmpresaService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.model.Empresa;
import br.com.portalproposta.backend.model.Validation;
import br.com.portalproposta.backend.repository.EmpresaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class EmpresaService {

    private final EmpresaRepository empresaRepository;

    @Autowired
    public EmpresaService(EmpresaRepository empresaRepository) {
        this.empresaRepository = empresaRepository;
    }

    public List<Empresa> list() {
        return empresaRepository.findAll();
    }

    public Empresa getById(String id) {
        return empresaRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("empresa parceira não encontrada"));
    }

    @Transactional
    public Empresa create(Empresa company) {
        if (company.getNome() == null || company.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do empresa é obrigatório");
        }

        if (company.getEmail() == null || company.getEmail().trim().isEmpty() || !company.getEmail().contains("@")) {
            throw new AppBadRequestException("email da empresa é obrigatório ou inválido");
        }

        company.setCnpj(Validation.cleanCPFCNPJ(company.getCnpj()));
        if (!Validation.isValidCNPJ(company.getCnpj())) {
            throw new AppBadRequestException("CNPJ inválido");
        }

        empresaRepository.findByCnpj(company.getCnpj()).ifPresent(existing -> {
            throw new AppBadRequestException("empresa com este CNPJ já cadastrada");
        });

        company.setId(UUID.randomUUID().toString());
        company.setAtivo(true);
        company.setCreatedAt(LocalDateTime.now());
        company.setUpdatedAt(LocalDateTime.now());

        return empresaRepository.save(company);
    }

    @Transactional
    public Empresa update(String id, Empresa input) {
        Empresa company = getById(id);

        if (input.getNome() == null || input.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome da empresa é obrigatório");
        }

        if (input.getEmail() == null || input.getEmail().trim().isEmpty() || !input.getEmail().contains("@")) {
            throw new AppBadRequestException("email da empresa é obrigatório ou inválido");
        }

        String cleanCNPJ = Validation.cleanCPFCNPJ(input.getCnpj());
        if (!cleanCNPJ.equals(company.getCnpj())) {
            if (!Validation.isValidCNPJ(cleanCNPJ)) {
                throw new AppBadRequestException("CNPJ inválido");
            }

            empresaRepository.findByCnpj(cleanCNPJ).ifPresent(existing -> {
                if (!existing.getId().equals(company.getId())) {
                    throw new AppBadRequestException("empresa com este CNPJ já cadastrada");
                }
            });
            company.setCnpj(cleanCNPJ);
        }

        company.setNome(input.getNome());
        company.setEmail(input.getEmail());
        company.setTelefone(input.getTelefone());
        company.setResponsavelNome(input.getResponsavelNome());
        company.setResponsavelEmail(input.getResponsavelEmail());
        company.setResponsavelTelefone(input.getResponsavelTelefone());
        company.setAtivo(input.isAtivo());
        company.setUpdatedAt(LocalDateTime.now());

        return empresaRepository.save(company);
    }

    @Transactional
    public void delete(String id) {
        Empresa company = getById(id);
        empresaRepository.delete(company);
    }
}

File: backend/Services/EmpresaService.cs
using Microsoft.EntityFrameworkCore;
using PortalProposta.Data;
using PortalProposta.Models;

namespace PortalProposta.Services;

public class EmpresaService(AppDbContext context)
{
    public async Task<List<Empresa>> ListAsync() =>
        await context.Empresas.ToListAsync();

    public async Task<Empresa> GetByIdAsync(string id) =>
        await context.Empresas.FirstOrDefaultAsync(e => e.Id == id)
            ?? throw new AppNotFoundException("empresa parceira não encontrada");

    public async Task<Empresa> CreateAsync(Empresa company)
    {
        Validate.Empresa(company);

        company.CNPJ = Validation.CleanCPFCNPJ(company.CNPJ);
        if (!Validation.IsValidCNPJ(company.CNPJ))
            throw new AppBadRequestException("CNPJ inválido");

        if (await context.Empresas.AnyAsync(e => e.CNPJ == company.CNPJ))
            throw new AppBadRequestException("empresa com este CNPJ já cadastrada");

        company.Id = Guid.NewGuid().ToString();
        company.Ativo = true;
        company.CreatedAt = company.UpdatedAt = DateTime.Now;

        context.Empresas.Add(company);
        await context.SaveChangesAsync();
        return company;
    }

    public async Task<Empresa> UpdateAsync(string id, Empresa input)
    {
        var company = await context.Empresas.FirstOrDefaultAsync(e => e.Id == id)
            ?? throw new AppNotFoundException("empresa parceira não encontrada");

        Validate.Empresa(input);

        var cleanCNPJ = Validation.CleanCPFCNPJ(input.CNPJ);
        if (cleanCNPJ != company.CNPJ)
        {
            if (!Validation.IsValidCNPJ(cleanCNPJ))
                throw new AppBadRequestException("CNPJ inválido");

            if (await context.Empresas.AnyAsync(e => e.CNPJ == cleanCNPJ && e.Id != company.Id))
                throw new AppBadRequestException("empresa com este CNPJ já cadastrada");

            company.CNPJ = cleanCNPJ;
        }

        company.Nome = input.Nome;
        company.Email = input.Email;
        company.Telefone = input.Telefone;
        company.ResponsavelNome = input.ResponsavelNome;
        company.ResponsavelEmail = input.ResponsavelEmail;
        company.ResponsavelTelefone = input.ResponsavelTelefone;
        company.Ativo = input.Ativo;
        company.UpdatedAt = DateTime.Now;

        await context.SaveChangesAsync();
        return company;
    }

    public async Task DeleteAsync(string id)
    {
        var company = await context.Empresas.FirstOrDefaultAsync(e => e.Id == id)
            ?? throw new AppNotFoundException("empresa parceira não encontrada");

        context.Empresas.Remove(company);
        await context.SaveChangesAsync();
    }
}
